### DISCUSSION TITLE
Can I interface Julia code into modelica?

### DISCUSSION THREAD

**Discussion by Amuri44** [OpenModelica/OpenModelica#8567] [Q&A]
I have a model in Modelica and I need to insert into it a part that built by Julia language. Is there a way for that?

**Comment by AnHeuermann** [comment]
Check out `12.9.4 Annotations for External Libraries and Include Files` from  the [Modelica Specification 3.5](https://modelica.org/documents/MLS.pdf).

You can create a C interface `foo` that can call your Julia code and call it from a Modelica function
```modelica
function foo
  input Real x ;
  output Real y ;
external "C"
  annotation ( Library = "MyLib",
  Include = "# include \"MyHeader.h\"") ;
end foo;
```

**Reply by Amuri44** [reply]
Thank for your reply.
So, first I need to translate my Julia code to C code, then I can include it by the below code?:
```modelica
external "C"
  annotation ( Library = "MyLib",
  Include = "# include \"MyHeader.h\"") ;
```

**Reply by AnHeuermann** [reply]
Not translate it but create an interface or use the C function pointers from Julia directly.

It's a bit difficult though.
Check out [@cfunction](https://docs.julialang.org/en/v1/base/c/#Base.@cfunction)  or [embedding Julia](https://docs.julialang.org/en/v1/manual/embedding/) and the Julia discourse.

I would suggest to start small and try to create a library that can be called from Modelica. You could start with the example 12.9.5.
And when that is working try to create a Julia function that will do `foo` and create a C interface to `foo`.

**Reply by Amuri44** [reply]
Ok, Thank you very much for your guide. I will check.

**Reply by Amuri44** [reply]
@AnHeuermann I am reading the `cfunction` and `embedding Julia`.
On the other hand, can I generate a `dll` file from the Julia code and use it directly in modelica?

**Reply by AnHeuermann** [reply]
You should probably ask in the Julia discourse about Julia. But I guess https://github.com/JuliaLang/PackageCompiler.jl would be an option. Never tried it though.

**Reply by Amuri44** [reply]
@AnHeuermann 
So, basically modelica can read `dll` as mentioned in `12.9.4 Annotations for External Libraries and Include Files`, but the doubts are:
1- if Julia can generate a `dll`
2- if the format of this `dll` is compitable with the format that can be read by modelica, right?

**Reply by mahge** [reply]
@Amuri44 If you have the whole output message (the compilation commands included) please add it. 

You can also try with just `Library = "HiAmro"` instead of `Library = "HiAmro.dll"` in the annotations and let us know.

**Reply by mahge** [reply]
So omc is looking for your library in these locations:
```
C:/Program Files/OpenModelica1.18.0-64bit/lib/omlibrary/OpenEMTP
C:/Program Files/OpenModelica1.18.0-64bit/lib//omc
C:/Program Files/OpenModelica1.18.0-64bit/lib/
C:/Users/amroa/AppData/Roaming/.openmodelica/binaries/OpenEMTP
C:/Program Files/OpenModelica1.18.0-64bit/lib/omlibrary/OpenEMTP/Resources/source_julia/bin/mingw64
C:/Program Files/OpenModelica1.18.0-64bit/lib/omlibrary/OpenEMTP/Resources/source_julia/bin/win64
C:/Program Files/OpenModelica1.18.0-64bit/lib/omlibrary/OpenEMTP/Resources/source_julia/bin
```

If your library is indeed in one of these locations, you can try the latest nightly build of omc. The second error you encountered (Process crashed Simulation process failed. Exited with code 0xffffffffc0000135.) has been fixed last week in #8572.

**Reply by mahge** [reply]
Then I think you will need to update your omc or as a temporary workaround you can copy the dll in to the directory `C:/Users/amroa/AppData/Local/Temp/OpenModelica/OMEdit/OpenEMTP.TestJuliaPackage.ExternaljuliaLibraries/`.

**Reply by mahge** [reply]
> Do I need to uninstall my current version of openmodelica, then install the nightly build version?

Yes you need to uninstall the current version. You can get the nightly version from here:
https://build.openmodelica.org/omc/builds/windows/nightly-builds/64bit/

**Reply by Amuri44** [reply]
> Yes you need to uninstall the current version. You can get the nightly version from here:
> https://build.openmodelica.org/omc/builds/windows/nightly-builds/64bit/

So, I need to install the whole file or just `https://build.openmodelica.org/omc/builds/windows/nightly-builds/64bit/OpenModelica-latest.exe`?

**Reply by mahge** [reply]
Just download OpenModelica latest exe. It is the installer.

**Comment by mahge** [comment]
Sorry for the delay. It seems your error messages point to an invalid dll. `c0000139` is DLL entry point not found. I am not sure what the problem is exactly as it can be many different reasons. Maybe the name of the function in the DLL is not exactly `test`? If you can attach the DLL itself we can take a look.

**Reply by Amuri44** [reply]
- The entry points are located in the `julia_init.h` that I included. But, I am not sure if it needs to be called in the modelica model as well as in this example to call the dll (generated by julia) in C tool [Call julia dll from C Tool](https://github.com/simonbyrne/libcg/blob/master/main-c/main.c)
- The zip of the dll package is big so I divided it over 5 zip files. The main is [Case.zip](https://github.com/OpenModelica/OpenModelica/files/8112541/Case.zip). The other four are below and need to be put in the `bin` folder.

**Reply by Amuri44** [reply]
[Complement-1.zip](https://github.com/OpenModelica/OpenModelica/files/8112548/Complement-1.zip)

**Reply by Amuri44** [reply]
[Complement-2.zip](https://github.com/OpenModelica/OpenModelica/files/8112552/Complement-2.zip)

**Reply by Amuri44** [reply]
[Complement-3.zip](https://github.com/OpenModelica/OpenModelica/files/8112554/Complement-3.zip)

**Reply by Amuri44** [reply]
[Complement-4.zip](https://github.com/OpenModelica/OpenModelica/files/8112560/Complement-4.zip)

**Reply by mahge** [reply]
Your DLL (HiAmro.dll) depends on at least two additional DLLs. libjulia.dll and libjulia-internal.dll. I am not saying this will fix it, but you can try adding these two to the annotations
```mo
annotation(Library = {"HiAmro", "libjulia", "libjulia-internal"}, ....)
```
I will take a look here as soon as I can.

**Reply by Amuri44** [reply]
Thanks for your reply and I am waiting you please!
I have tried to add the libraries as `annotation(Library = {"HiAmro", "libjulia", "libjulia-internal"}, ....)` but I got the same error as above. I have just been informed recently that an entry-points function `init_julia()` located in `..\include\julia_init.h` has to be called before the main function `test()` such as in here [Call julia dll from C Tool](https://github.com/simonbyrne/libcg/blob/master/main-c/main.c). I could know how to call it correctly?

**Reply by mahge** [reply]
Probably the best path is to first see if you can successfully link to and call the function in julia generated DLL _from C code as_ the example shows. If you can manage that then you can make small C function that calls `test` from the julia DLL, compile that small file (containing the init_julia call and the wrapper to test) to a DLL, and then use call the function from the new DLL in OpenModelica following the same procedure. 

I am trying to see what more is needed to directly run your DLL. There seems to be more DLLs needed (e.g. LLVM.dll ...)

**Comment by mahge** [comment]
Unfortunately, I have to abandon this for now since it is basically a wild goose chase at this point. Even with all the DLLs available there is still entry point issues in the LLVM dll. It is getting outside of our support domain. Maybe the good people from Julia can point you in the right direction. 

I recommend you try to follow the guide and try to call the Julia function from C. If you successfully achieve that I can help you with adapting it to be used with Modelica. For now there are other more pressing issues. Don't give up though :). Good luck.

**Reply by mahge** [reply]
You can put it in another function, e.g. `ExternalJuliaInit` and call that function in `ExternaljuliaLibraries` before you call `Externaljuliafunc1`. however, I do not think it will make a difference as the problem is that the executable is failing to even start. So it is probably not an issue with `init_julia()` (at least not yet.)

**Comment by adrpo** [comment]
Call 
`julia julia-config.jl --libs` 
to see what is needed

**Reply by Amuri44** [reply]
@adrpo where should I put this code?

**Reply by adrpo** [reply]
Myself I call it from mingw64 terminal:
`/e/bin/julia/bin/julia.exe /e/bin/julia/share/julia/julia-config.jl --ldflags --ldlibs`
but you can use cmd as well:
`e:/bin/julia/bin/julia.exe e:/bin/julia/share/julia/julia-config.jl --ldflags --ldlibs`

However this does not help as it will only list this:
```
-L'E:\bin\julia\bin' -Wl,--stack,8388608
-ljulia -lopenlibm
```

And, at leat in my system, Julia library does not depend on anything (but the kernel):

```
ntldd.exe /e/bin/julia/bin/libjulia.dll
        KERNEL32.dll => C:\WINDOWS\SYSTEM32\KERNEL32.dll (0x0000000000ce0000)

```

What is in your HiAmro.dll, how did you build it?!

**Reply by adrpo** [reply]
Ok. Sounds good.

**Reply by adrpo** [reply]
julia-config.jl is not in the bin directory is in ../share/julia/ relative to the bin directory. 
But this doesn't matter, just skip it.

**Reply by Amuri44** [reply]
Ok

**Comment by adrpo** [comment]
I can have a quick look later as I have been playing with calling julia from C in our Modelica parser for Julia.

**Comment by adrpo** [comment]
These are the DLLs you need to run HiAmro.dll:

```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33
# ntldd -R /c/Users/adrpo33/Downloads/Case/bin/HiAmro.dll | grep -vi Windows | grep -v "ext-ms" | grep -v HvsiFileTrust.dll
        libgcc_s_seh-1.dll => C:\OMDev\tools\msys\mingw64\bin\libgcc_s_seh-1.dll (0x0000000000620000)
         libwinpthread-1.dll => C:\OMDev\tools\msys\mingw64\bin\libwinpthread-1.dll (0x00000000001f0000)
        libjulia-internal.dll => E:\bin\julia\bin\libjulia-internal.dll (0x00000000071a0000)
         libLLVM.dll => C:\OMDev\tools\msys\mingw64\bin\libLLVM.dll (0x000000000c2c0000)
          zlib1.dll => C:\OMDev\tools\msys\mingw64\bin\zlib1.dll (0x0000000000620000)
          libffi-7.dll => C:\OMDev\tools\msys\mingw64\bin\libffi-7.dll (0x00000000001f0000)
         libssp-0.dll => C:\Users\adrpo33\Downloads\Case\bin\libssp-0.dll (0x0000000000620000)
         libstdc++-6.dll => C:\OMDev\tools\msys\mingw64\bin\libstdc++-6.dll (0x0000000012010000)
        libjulia.dll => E:\bin\julia\bin\libjulia.dll (0x0000000000620000)
```

All the ones found in "C:\OMDev\tools\msys\mingw64" are actually from the OpenModelica environment, which could be problematic as they might be different versions, etc. Where did you get the other dlls from?

As far as I can tell, if you copy *all* the dlls from your julia\bin to where your HiAmro.dll is it might just work. Try that and if is not working,  I will do some more investigation.

**Reply by Amuri44** [reply]
> All the ones found in "C:\OMDev\tools\msys\mingw64" are actually from the OpenModelica environment, which could be problematic as they might be different versions, etc. Where did you get the other dlls from?
> 
> As far as I can tell, if you copy _all_ the dlls from your julia\bin to where your HiAmro.dll is it might just work. Try that and if is not working, I will do some more investigation.

When I built my library `HiAmro`, the generated files are in two folders `..\bin` (which include my library and all other dlls are automatically existed by this it can release me from copying them from julia\bin) and `..\include`(which include `julia_init.h`)

**Comment by adrpo** [comment]
As far as I can see test is defined in your dll, is the last one:
```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33
# nm /c/Users/adrpo33/Downloads/Case/bin/HiAmro.dll | grep test
0000000070aa3710 I __imp_jl_f__call_latest
000000006ae822b0 t japi1_membershiptest_25276
000000006b12f900 t jfptr_already_inferred_quick_test_8714
000000006af8af40 t jfptr_writeshortest_21348
000000006b1b74a8 T jl_f__call_latest
0000000070a9de20 b jl_symYY.invokelatest4725
000000006ae813b0 t jlcapi_test_15340
000000006b12f8f0 t julia_already_inferred_quick_test_8713
000000006af88880 t julia_writeshortest_21347
0000000070a93328 b MUL.CoreDOT._call_latest3002
0000000070a973d0 b MUL.LinearAlgebraDOT.BLASDOT._testmat6737
0000000070a97800 b MUL.MainDOT.BaseDOT.invokelatest4083
0000000070a98fd8 b MUL.MainDOT.BaseDOT.RyuDOT.writeshortest4257
0000000070a9b860 b SUM.PkgDOT.APIDOT.YY.test2788
000000006ae813b0 T test
```

**Reply by adrpo** [reply]
Ouch,  I meant to copy them to where the dll on the disk is, not list them in the Modelica code. In the Modelica code list just HiAmro, I think it should be enough.

**Reply by adrpo** [reply]
Ok. I will try to debug it. It should just work, but there are always details that we can forget about.

**Comment by adrpo** [comment]
Ok, it seems there are some conflicts between the dlls used by HiAmro.dll and the dlls we have in OpenModelica. If I put the HiAmro dlls in the path last then I get `c0000139 is DLL entry point not found`, if I put them first then I get a crash when test is called:
```
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 --
```

I tried to debug it but that didn't help too much either. I put a breakpoint on `init_julia` and that one doesn't seem to be called. 

This is my code (made the init_julia call an initial equation to be called once):
```modelica
within OpenEMTP;

package TestJuliaPackage
  //extends UserGuide.Icons.Library;

  function Externaljuliaini
    //extends Modelica.Icons.Function;
    input Real t;
    external
    init_julia(0, 0) annotation(Include="#include \"julia_init.h\"", IncludeDirectory="modelica://OpenEMTP/Resources/source_julia/include",Library = {"HiAmro", "libjulia", "libjulia-internal"},LibraryDirectory="modelica://OpenEMTP/Resources/source_julia/bin");
  
  end Externaljuliaini;
  
  function Externaljuliafunc1
    //extends Modelica.Icons.Function;
    output Integer y;
    external
    y=test() annotation(Include="#include \"julia_init.h\"", IncludeDirectory="modelica://OpenEMTP/Resources/source_julia/include",Library = {"HiAmro", "libjulia", "libjulia-internal"},LibraryDirectory="modelica://OpenEMTP/Resources/source_julia/bin");
  
  end Externaljuliafunc1;

  model ExternaljuliaLibraries
    Integer y;
  initial equation
   Externaljuliaini(time);
  equation
   y = Externaljuliafunc1();
  end ExternaljuliaLibraries;
end TestJuliaPackage;
```

The args of `init_julia` are: ` init_julia(int argc, char *argv[])` so calling it with 0, 0 should work, but maybe not. I will try a different approach later.
I guess the best would be to link the HiAmro.dll to a static Julia library so we don't get the dll conflicts but I don't know if that's possible.
I will try to play with this more in the next days and see if I can find a solution.

**Reply by Amuri44** [reply]
Thank you very much for the details to understand the causing issue.
> The args of `init_julia` are: ` init_julia(int argc, char *argv[])` so calling it with 0, 0 should work, but maybe not.
I run it but gave me `Process crashed Simulation process failed. Exited with code 0xffffffffc0000139.`
> I will try to play with this more in the next days and see if I can find a solution.
Thanks for your contribution!

**Reply by Amuri44** [reply]
---I will try to play with this more in the next days and see if I can find a solution.
@adrpo Hello again and I hope you are doing well. Any new update for this case?

**Comment by adrpo** [comment]
I did not had time to look into this yet. Are you bound to Windows? I think this would work fine in Linux.

But on Windows is a dll hell, we have the dlls in OpenModelica/tools/msys/mingw64/bin and lib which we link the simulation executables with but then we also link with your dll which was build using the Julia dlls which are different versions. 
The only way this might work is 
- to use the same mingw in OpenModelica as the ones in Julia but that would never happen.
- to somehow have your dll linked statically to the Julia lib but I don't think there is any static Julia lib on Windows, just a dll.

**Reply by Amuri44** [reply]
@adrpo thanks to your reply. Yes, I am bound to Windows.

> But on Windows is a dll hell, we have the dlls in OpenModelica/tools/msys/mingw64/bin and lib which we link the simulation executables with but then we also link with your dll which was build using the Julia dlls which are different versions.

I understand the problem more now.

> * to use the same mingw in OpenModelica as the ones in Julia but that would never happen

Why?

> * to somehow have your dll linked statically to the Julia lib but I don't think there is any static Julia lib on Windows, just a dll.

I seared on this, but yes I didtt see any static julia lib

**Comment by adrpo** [comment]
Another way to fix this is to use some tcp/ip or some other way of communication between the simulation executable and your Julia based process.

**Reply by Amuri44** [reply]
Could you please give me a link or example to to this as I dint work on it before?
